### ISSUE TITLE
Bug: no-floating-promises should ignore operations that return the promise passed in

### ISSUE THREAD

**Issue by mqudsi** [typescript-eslint/typescript-eslint#7579]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=GYVwdgxgLglg9mABMOcAUBKAXIgCgJzgFsYBnAUwB4A3OGAEwD5EBvAWAChFEAbcqRAAdCJCogC8iMOQDueEWSq0GjNJgnMWiAL4YA3J24B5AEYArctAB09csBjSCcQeXxQAnmmHFFAGkQAyu5EJnA8VlBwAVD4DgDmACoAhnH%2B7FzciHH86umZmfj8IPhIAERJQgoUpQYZ3NqGOvqNhVDFSN6i5LXaQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
function foo(): Promise<void> {
  let promise = new Promise<void>(() => { });
  // Contrived use of Object.defineProperty() for illustration purposes:
  Object.defineProperty(promise, Symbol.toStringTag, {
    get() {
      return "a promise";
    }
  });
  return promise;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  "rules": {
    "@typescript-eslint/no-floating-promises": "error"
  },
};
```

### Expected Result

The return value of functions like Object.defineProperty and Object.assign should be ignored by no-floating-promises *iff* the promise passed in as the source or target was already bound to a variable, since there is no new promise - just another binding to the same old one (which has already been linted or will be linted separately).

e.g.

this should not error:

```ts
declare foo: Promise;

Object.assign(foo, { ... });
```

because there is no new promise created as a result of the call to `Object.assign()` (or the call to `Object.defineProperty()` in the example above).

but this should:

```ts
Object.assign(new Promise(() => {}), { ... });
```

because there is no existing binding to the promise that was passed in to `Object.assign()`.

### Actual Result

eslint-typescript is unable to realize that there is no new promise being created as a result of the call to `Object.defineProperty()` or `Object.assign()` and complains that the return value of the Object.foo call is not being awaited/caught when it very well might be since there's another binding to the promise.

**Comment by bradzacher** [comment]
There's no way for us to know what the function call is doing.
Your assertion does not hold for all functions.

For example:

```ts
function foo(arg: Promise<void>) {
  await arg;
  throw new Error();
}


const prom = new Promise();
foo(prom);
```

There's no way for us to determine how the function actually uses the promise and so ignoring this case will lead to false negatives with floating promises.

**Comment by mqudsi** [comment]
I did not mean for user-provided functions but only for well-known functions in the standard library or the DOM.
